hls4ml/templates/vivado/nnet_utils/nnet_dense_resource.h

```diff
@@ -261,6 +261,247 @@ void dense_resource_rf_gt_nin(
     }
 }
 
+// Dense (with ReLU)
+template<class data_T, class res_T, typename CONFIG_T>
+void dense_relu_resource_rf_leq_nin(
```

Comment:
this seems like a lot of code duplication. Is it not better to just check `CONFIG_T::merged_relu` inside the `dense_resource_rf_leq_nin` to see whether or not to apply the ReLU in the existing function? Or does that result in a performance penalty for some reason?

Reply:
Unclear if it would incur a performance penalty. We felt like separating out the functions as `dense()` and `dense_relu()` would make it more clear when the dense-relu merging was happening based on the function calls, even though this does duplicate code.  

We can test if it will affect resource utilization.

Reply:
We tested checking `CONFIG_T::merged_relu` inside the `dense_resource*()` functions, and it costs more resources to do it that way. 

| RN07                                                           | BRAMs (%) | DSPs (%) | FFs(%) | LUTs (%) |
|----------------------------------------------------------------|:---------:|:--------:|:------:|:--------:|
| Non-merged (default hls4ml)                                    |    142    |     0    |   64   |    113   |
| Non-merged  (merged_relu optimizer off - compact code version) |    166    |     0    |   69   |    115   |
| Merged (merged_relu optimizer on - compact code version)       |    119    |     0    |   63   |    93    |
| **Merged (verbose code version - CURRENT)**                      |     95    |     0    |   58   |    91    |